File: TrinhsGroup/View/Profile/OrderStatusPresentation.swift
//
//  OrderStatusPresentation.swift
//  TrinhsGroup
//
//  Single source of truth for how a WooCommerce order status is shown to the customer.
//  Copy, icon, colour and stage all live here so the hero, the stepper and anything added
//  later cannot drift into describing the same order differently.
//

import SwiftUI

// MARK: - Stage

/// The four steps of a normal order, in order.
///
/// A stage is *reached* by any status at or past it: an order being cooked was obviously
/// confirmed first, so `processing` reaches `.received` as well as `.cooking`.
enum OrderStage: Int, CaseIterable, Comparable {
    case placed = 0
    case received
    case cooking
    case ready

    static func < (lhs: OrderStage, rhs: OrderStage) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    var title: String {
        switch self {
        case .placed:   return "Order Placed"
        case .received: return "Order Received"
        case .cooking:  return "In the Kitchen"
        case .ready:    return "Ready for Pickup"
        }
    }

    var icon: String {
        switch self {
        case .placed:   return "bag.fill"
        case .received: return "bell.badge.fill"
        case .cooking:  return "flame.fill"
        case .ready:    return "checkmark.seal.fill"
        }
    }

    /// Colour of this stage's node once it has been completed.
    ///
    /// Fixed per stage rather than inherited from the order's *current* status. Inheriting
    /// meant a cancelled order painted the stages it genuinely completed in failure red,
    /// as if they had failed too. As a bonus the rail now warms from coral through orange
    /// to green as the food progresses.
    var tint: Color {
        switch self {
        case .placed, .received: return StatusPalette.brand
        case .cooking:           return StatusPalette.cooking
        case .ready:             return StatusPalette.ready
        }
    }

    /// The stage a status implies, or `nil` for statuses that end the order early.
    ///
    /// `placed` is the synthetic status the server uses to timestamp order creation;
    /// `pending` sits at the same stage because an unpaid order has been placed but not
    /// yet confirmed by anyone.
    init?(status: String) {
        switch status {
        case "placed", "pending": self = .placed
        case "on-hold":           self = .received
        case "processing":        self = .cooking
        case "completed":         self = .ready
        default:                  return nil
        }
    }
}

// MARK: - Palette

/// Warm ramp that tracks the food: amber while waiting, orange in the kitchen, green when
/// ready. Explicit hex rather than the asset catalogue because this screen is drawn on a
/// hardcoded `f9f9f9` background and `ColorPrimary`'s dark variant is grey.
private enum StatusPalette {
    static let waiting  = Color(hex: "E9A23B")
    static let brand    = Color(hex: "FE7058")   // matches ColorPrimary's light value
    static let cooking  = Color(hex: "F2682A")
    static let ready    = Color(hex: "57A733")
    static let failure  = Color(hex: "E5484D")
}

// MARK: - Presentation

/// Everything the UI needs to render one order status.
struct OrderStatusPresentation {

    let title: String
    /// One line telling the customer what is happening, in their words not WooCommerce's.
    let subtitle: String
    let icon: String
    /// Name of a bundled Lottie animation to use instead of `icon`, when one fits.
    let lottieName: String?
    let tint: Color
    /// `true` for statuses that end the order. Suppresses the pulse, not the Lottie —
    /// `Order_refunded` and `Order_failed` were authored for these states.
    let isTerminal: Bool

    init(status: String) {
        switch status {
        case "pending":
            title      = "Awaiting Payment"
            subtitle   = "Finish paying to send your order to the kitchen."
            icon       = "creditcard.fill"
            lottieName = nil
            tint       = StatusPalette.waiting
            isTerminal = false

        case "on-hold":
            title      = "Order Received"
            subtitle   = "We're confirming your order with the kitchen."
            icon       = "bell.badge.fill"
            lottieName = "Order_onHold"
            tint       = StatusPalette.brand
            isTerminal = false

        case "processing":
            title      = "In the Kitchen"
            subtitle   = "Our chefs are cooking your food right now."
            icon       = "flame.fill"
            // Note: `cookLoading.json`, despite its name, is a *text* animation — five text
            // layers spelling "Cooking...." with bouncing dots, built for the 200pt loading
            // overlay. Order_processing is the purpose-made one.
            lottieName = "Order_processing"
            tint       = StatusPalette.cooking
            isTerminal = false

        case "completed":
            title      = "Ready for Pickup"
            subtitle   = "Your order is ready — enjoy your meal!"
            icon       = "checkmark.seal.fill"
            lottieName = "Order_ready"
            tint       = StatusPalette.ready
            isTerminal = false

        case "cancelled":
            title      = "Order Cancelled"
            subtitle   = "This order was cancelled."
            icon       = "xmark.circle.fill"
            // No Order_cancelled.json was supplied, and Order_failed is not a stand-in:
            // a customer cancelling their own order is not a payment failure.
            lottieName = nil
            tint       = StatusPalette.failure
            isTerminal = true

        case "refunded":
            title      = "Order Refunded"
            subtitle   = "Your payment has been returned."
            icon       = "arrow.uturn.backward.circle.fill"
            lottieName = "Order_refunded"
            tint       = StatusPalette.waiting
            isTerminal = true

        case "failed":
            title      = "Payment Failed"
            subtitle   = "The payment didn't go through."
            icon       = "exclamationmark.triangle.fill"
            lottieName = "Order_failed"
            tint       = StatusPalette.failure
            isTerminal = true

        default:
            // A status WooCommerce has but the app doesn't model — a custom one added in
            // wp-admin, say. Show the slug rather than pretending to know what it means.
            title      = status.replacingOccurrences(of: "-", with: " ").capitalized
            subtitle   = "We'll update you as your order progresses."
            icon       = "clock.fill"
            lottieName = nil
            tint       = StatusPalette.brand
            isTerminal = false
        }
    }
}

// MARK: - Stepper node

/// How one node of the four-step rail is drawn.
enum OrderStepState {
    case done
    case current
    case upcoming
}

/// One rendered row of the progress rail.
struct OrderStep: Identifiable {
    let id: Int
    let title: String
    let icon: String
    let state: OrderStepState
    /// Pre-formatted for display, or `nil` when this step has no known timestamp.
    let timestamp: String?
    /// Colour for this node and for the rail leading into it. Resolved here so the view
    /// never has to reason about which of several tints applies.
    let tint: Color
    /// `true` for the appended cancelled / refunded / failed node.
    let isTerminal: Bool
}

/// Reduces an order plus its status history into the rail's rows.
///
/// Pure and free of SwiftUI state so it can be tested directly.
enum OrderProgressBuilder {

    /// - Parameters:
    ///   - status: the order's current WooCommerce status.
    ///   - events: status history, oldest first. May be empty — see `fallbackEvents`.
    static func steps(status: String, events: [OrderTimelineEvent]) -> [OrderStep] {
        let presentation = OrderStatusPresentation(status: status)

        // The furthest stage the order actually got to. For a terminal status this is the
        // last happy-path stage seen in the history, which is why the history matters:
        // "cancelled" alone cannot say whether the food had started cooking.
        let reached = events.compactMap { OrderStage(status: $0.status) }.max()

        // Earliest timestamp recorded for each stage.
        var stamps: [OrderStage: String] = [:]
        for event in events {
            guard let stage = OrderStage(status: event.status) else { continue }
            if stamps[stage] == nil { stamps[stage] = event.displayTime }
        }

        let current = presentation.isTerminal ? nil : OrderStage(status: status)

        var steps: [OrderStep] = OrderStage.allCases.compactMap { stage in
            // A terminal order's rail stops at the stage it reached; the nodes it never
            // got to are replaced by the terminal node rather than shown as pending.
            if presentation.isTerminal, let reached, stage > reached { return nil }

            let state: OrderStepState
            if let current {
                if stage < current      { state = .done }
                else if stage == current { state = .current }
                else                     { state = .upcoming }
            } else {
                state = .done   // terminal: everything it reached is behind it
            }

            return OrderStep(
                id: stage.rawValue,
                title: stage.title,
                icon: stage.icon,
                state: state,
                timestamp: stamps[stage],
                // The node the order is sitting on takes the hero's colour so the two
                // agree; completed nodes keep their own stage colour.
                tint: state == .current ? presentation.tint : stage.tint,
                isTerminal: false
            )
        }

        if presentation.isTerminal {
            steps.append(OrderStep(
                id: OrderStage.allCases.count,
                title: presentation.title,
                icon: presentation.icon,
                state: .current,
                timestamp: events.last(where: { $0.status == status })?.displayTime,
                tint: presentation.tint,
                isTerminal: true
            ))
        }

        return steps
    }

    /// Timeline to use when the history endpoint gave us nothing — it isn't deployed yet,
    /// the request failed, or the order predates it.
    ///
    /// Reproduces what the screen showed before this endpoint existed: creation time on
    /// the first step, last-modified on the current one.
    ///
    /// See `orderTimelineStamp` for the timezone caveat these two dates inherit.
    static func fallbackEvents(for order: Order) -> [OrderTimelineEvent] {
        var events: [OrderTimelineEvent] = []

        if !order.dateCreated.isEmpty {
            events.append(OrderTimelineEvent(
                status: "placed",
                displayTime: order.dateCreated.orderTimelineStamp
            ))
        }

        if order.status != "placed", !order.dateModified.isEmpty {
            events.append(OrderTimelineEvent(
                status: order.status,
                displayTime: order.dateModified.orderTimelineStamp
            ))
        }

        return events
    }
}

File: kmp/shared/src/commonMain/kotlin/com/trinhsgroup/shared/order/OrderStatusPresentation.kt
package com.trinhsgroup.shared.order

import com.trinhsgroup.shared.model.Order
import com.trinhsgroup.shared.model.OrderTimelineEvent
import com.trinhsgroup.shared.model.orderTimelineStamp

/**
 * Single source of truth for how a WooCommerce status is shown to a customer.
 * Mirrors Swift's OrderStatusPresentation.swift.
 *
 * Copy, icon, colour and stage all live here so the hero, the stepper, and anything added later
 * cannot drift into describing the same order differently. Colours are named rather than given
 * as values — the platform layer maps a [StatusTint] to its own palette.
 */

/**
 * The four steps of a normal order, in order.
 *
 * A stage is *reached* when a status implies it; statuses that end an order early have no stage.
 */
enum class OrderStage(val order: Int, val title: String) {
    PLACED(0, "Order Placed"),
    RECEIVED(1, "Order Received"),
    COOKING(2, "In Kitchen"),
    READY(3, "Ready for Pickup");

    /** Colour of this stage's node once it has been completed. */
    val tint: StatusTint
        get() = when (this) {
            PLACED, RECEIVED -> StatusTint.BRAND
            COOKING -> StatusTint.COOKING
            READY -> StatusTint.READY
        }

    val icon: StatusIcon
        get() = when (this) {
            PLACED -> StatusIcon.BAG
            RECEIVED -> StatusIcon.BELL
            COOKING -> StatusIcon.FLAME
            READY -> StatusIcon.SEAL
        }

    companion object {
        /**
         * The stage a status implies, or null for statuses that end an order early.
         *
         * `placed` is the synthetic status the server uses for the timestamp of order creation;
         * `pending` sits in the same stage — an unpaid order has been placed but not yet
         * confirmed by anyone.
         */
        fun of(status: String): OrderStage? = when (status) {
            "placed", "pending" -> PLACED
            "on-hold" -> RECEIVED
            "processing" -> COOKING
            "completed" -> READY
            else -> null
        }
    }
}

/**
 * Warm ramp that tracks the food: amber while waiting, orange in the kitchen, green when ready.
 * Named so the shared layer holds no platform colour type.
 */
enum class StatusTint { WAITING, BRAND, COOKING, READY, FAILURE }

enum class StatusIcon { BAG, BELL, FLAME, SEAL, CARD, CROSS, REFUND, WARNING, CLOCK }

/** What to show for one status: the customer's words, not WooCommerce's. */
data class OrderStatusPresentation(
    val title: String,
    val subtitle: String,
    val icon: StatusIcon,
    /** Name of a bundled Lottie animation to use instead of [icon], when one fits. */
    val lottieName: String?,
    val tint: StatusTint,
    /** True when the status ends the order — nothing further is coming. */
    val isTerminal: Boolean
) {
    companion object {
        fun of(status: String): OrderStatusPresentation = when (status) {
            "pending" -> OrderStatusPresentation(
                title = "Awaiting Payment",
                subtitle = "We'll start once your payment goes through.",
                icon = StatusIcon.CARD,
                lottieName = null,
                tint = StatusTint.WAITING,
                isTerminal = false
            )

            "on-hold" -> OrderStatusPresentation(
                title = "Order Received",
                subtitle = "We're confirming your order with the kitchen.",
                icon = StatusIcon.BELL,
                lottieName = "Order_onHold",
                tint = StatusTint.BRAND,
                isTerminal = false
            )

            "processing" -> OrderStatusPresentation(
                title = "In Kitchen",
                subtitle = "Our chefs are cooking your food right now.",
                icon = StatusIcon.FLAME,
                lottieName = "Order_processing",
                tint = StatusTint.COOKING,
                isTerminal = false
            )

            "completed" -> OrderStatusPresentation(
                title = "Ready for Pickup",
                subtitle = "Your order is ready — enjoy your meal!",
                icon = StatusIcon.SEAL,
                lottieName = "Order_ready",
                tint = StatusTint.READY,
                isTerminal = false
            )

            "cancelled" -> OrderStatusPresentation(
                title = "Order Cancelled",
                subtitle = "This order was cancelled.",
                icon = StatusIcon.CROSS,
                lottieName = null,
                tint = StatusTint.FAILURE,
                isTerminal = true
            )

            "refunded" -> OrderStatusPresentation(
                title = "Order Refunded",
                subtitle = "This order was refunded.",
                icon = StatusIcon.REFUND,
                lottieName = "Order_refunded",
                tint = StatusTint.WAITING,
                isTerminal = true
            )

            "failed" -> OrderStatusPresentation(
                title = "Payment Failed",
                subtitle = "The payment didn't go through.",
                icon = StatusIcon.WARNING,
                lottieName = "Order_failed",
                tint = StatusTint.FAILURE,
                isTerminal = true
            )

            else -> OrderStatusPresentation(
                // A status WooCommerce has but the app doesn't model — a custom one added in
                // wp-admin, say. Show the slug rather than pretending to know what it means.
                title = status.replace("-", " ").split(" ")
                    .joinToString(" ") { word -> word.replaceFirstChar { it.uppercase() } },
                subtitle = "",
                icon = StatusIcon.CLOCK,
                lottieName = null,
                tint = StatusTint.BRAND,
                isTerminal = false
            )
        }
    }
}

enum class OrderStepState { DONE, CURRENT, UPCOMING }

/** One row of the progress rail. */
data class OrderStep(
    val id: Int,
    val title: String,
    val icon: StatusIcon,
    val state: OrderStepState,
    /** Pre-formatted for display, or null when the step has no known timestamp. */
    val timestamp: String?,
    val tint: StatusTint,
    /** True for the appended cancelled / refunded / failed node. */
    val isTerminal: Boolean
)

/**
 * Reduces an order plus its status history into the rail's rows.
 *
 * Pure and free of UI state so it can be tested directly.
 */
object OrderProgressBuilder {

    /**
     * @param status the order's current WooCommerce status
     * @param events status history, oldest first. May be empty — see [fallbackEvents].
     */
    fun steps(status: String, events: List<OrderTimelineEvent>): List<OrderStep> {
        val presentation = OrderStatusPresentation.of(status)

        // How far the order actually got. For a terminal status this is what the happy path
        // cannot say: "cancelled" alone cannot tell you whether the food had started cooking.
        val reached = events.mapNotNull { OrderStage.of(it.status) }.maxByOrNull { it.order }

        // Earliest timestamp recorded for each stage.
        val stamps = mutableMapOf<OrderStage, String?>()
        for (event in events) {
            val stage = OrderStage.of(event.status) ?: continue
            if (!stamps.containsKey(stage)) stamps[stage] = event.displayTime
        }

        val current = if (presentation.isTerminal) null else OrderStage.of(status)

        val steps = OrderStage.entries.mapNotNull { stage ->
            // A terminal order's rail stops at the stage it reached; nodes it never got to are
            // replaced by the terminal node rather than shown as pending.
            if (presentation.isTerminal && reached != null && stage.order > reached.order) {
                return@mapNotNull null
            }

            val state = when {
                current == null -> OrderStepState.DONE // terminal: all it reached is behind it
                stage.order < current.order -> OrderStepState.DONE
                stage.order == current.order -> OrderStepState.CURRENT
                else -> OrderStepState.UPCOMING
            }

            OrderStep(
                id = stage.order,
                title = stage.title,
                icon = stage.icon,
                state = state,
                timestamp = stamps[stage],
                // The node the order sits on takes the hero's colour so the two agree;
                // completed nodes keep their own stage colour.
                tint = if (state == OrderStepState.CURRENT) presentation.tint else stage.tint,
                isTerminal = false
            )
        }.toMutableList()

        if (presentation.isTerminal) {
            steps.add(
                OrderStep(
                    id = OrderStage.entries.size,
                    title = presentation.title,
                    icon = presentation.icon,
                    state = OrderStepState.CURRENT,
                    timestamp = events.lastOrNull { it.status == status }?.displayTime,
                    tint = presentation.tint,
                    isTerminal = true
                )
            )
        }

        return steps
    }

    /**
     * Timeline to use when the history endpoint gave us nothing — the request failed, or the
     * order predates it.
     *
     * Shows creation time on the first step and last-modified on the current one.
     */
    fun fallbackEvents(order: Order): List<OrderTimelineEvent> {
        val events = mutableListOf<OrderTimelineEvent>()

        if (order.dateCreated.isNotEmpty()) {
            events.add(OrderTimelineEvent("placed", orderTimelineStamp(order.dateCreated)))
        }

        if (order.status != "placed" && order.dateModified.isNotEmpty()) {
            events.add(OrderTimelineEvent(order.status, orderTimelineStamp(order.dateModified)))
        }

        return events
    }
}
